Sources/Client.swift

```diff
@@ -0,0 +1,334 @@
+        let payload = JSONRPC_2_0.Request(method: "eth_sendRawTransaction",
+                                          params: .positional([.string(data)]),
+                                          id: .string(UUID().uuidString))
+        let request = Request(payload: payload, url: url)
+        try send(request, completion: completion)
+    }
+
+    override func onConnect(to url: WCURL) {
+        print("WC: client didConnect url: \(url.bridgeURL.absoluteString)")
+        if let session = communicator.session(by: url) { // reconnecting existing session
+            communicator.subscribe(on: session.dAppInfo.peerId, url: session.url)
+            delegate.client(self, didConnect: session)
+        } else { // establishing new connection, handshake in process
+            communicator.subscribe(on: dAppInfo.peerId, url: url)
+            let requestID = nextRequestId()
+            let createRequest = try! CreateSessionRequest(url: url, dAppInfo: dAppInfo, id: requestID)!
+            responses.add(requestID: requestID) { [unowned self] response in
+                self.handleHandshakeResponse(response)
+            }
+            communicator.send(createRequest, topic: url.topic)
+        }
+    }
+
+    private func nextRequestId() -> JSONRPC_2_0.IDType {
+        return JSONRPC_2_0.IDType.int(UUID().hashValue)
+    }
+
+    private func handleHandshakeResponse(_ response: Response) {
+        guard let session = try? Session(wcSessionResponse: response, dAppInfo: dAppInfo),
+            session.walletInfo!.approved else {
+                delegate.client(self, didFailToConnect: response.url)
+                return
+        }
+        communicator.addSession(session)
+        delegate.client(self, didConnect: session)
+    }
+
+    override func onTextReceive(_ text: String, from url: WCURL) {
+        if let response = try? communicator.response(from: text, url: url) {
+            log(response)
+            if let completion = responses.find(requestID: response.payload.id) {
+                completion(response)
+                responses.remove(requestID: response.payload.id)
+            }
+        } else if let request = try? communicator.request(from: text, url: url) {
+            log(request)
+            expectUpdateSessionRequest(request)
+        }
+    }
+
+    private func expectUpdateSessionRequest(_ request: Request) {
+        if request.payload.method == "wc_sessionUpdate" {
+            guard let params = request.payload.params,
+                case JSONRPC_2_0.Request.Params.positional(let arrayWrapper) = params, !arrayWrapper.isEmpty,
+                case JSONRPC_2_0.ValueType.object(let sessionUpdateParams) = arrayWrapper[0],
+                let requiredApproved = sessionUpdateParams["approved"],
+                case JSONRPC_2_0.ValueType.bool(let approved) = requiredApproved else {
+                    try! send(Response(payload: JSONRPC_2_0.Response.invalidJSON, url: request.url))
```

Comment:
This unwrapping response is hard to read, maybe extract it to a helper routine?

Reply:
ok